### PR TITLE
[Refactoring] refactor cluster reconciling defer

### PR DESCRIPTION
**Pull request by CiraciNicolo**
# Summary

Refactor the cluster reconcile defer to better handle errors.

# Additional Info

This PR is a cherry pick of #4282 as requested by @leonardoce

### CODE REVIEW

File: controllers/cluster_controller.go [hunk 1]
```diff
@@ -123,13 +123,17 @@ var ErrNextLoop = utils.ErrNextLoop
 // Reconcile is the operator reconcile loop
 func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	contextLogger, ctx := log.SetupLogger(ctx)
-
-	contextLogger.Debug(fmt.Sprintf("reconciling object %#q", req.NamespacedName))
+	var err error
 
 	defer func() {
-		contextLogger.Debug(fmt.Sprintf("object %#q has been reconciled", req.NamespacedName))
+		if err != nil {
```
Comment (leonardoce) [inline controllers/cluster_controller.go:129]: Errors are already reported here: https://github.com/kubernetes-sigs/controller-runtime/blob/9bc967a711e9538cff4099645dd6df64c9ba8a74/pkg/internal/controller/controller.go#L324

We don't need to report them another time.
Here, we need something to help us to know when a reconciliation loop started and when it finished.
Let me try proposing something else.

**Comment by leonardoce** [comment]
With the proposed implementation, the logs will look like:

```
{"level":"debug","ts":"2024-05-02T14:49:10Z","msg":"Reconciliation loop start","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-example","namespace":"default"},"namespace":"default","name":"cluster-example","reconcileID":"3f432bf6-4f9c-4610-a510-6c097d46f98d","caller":"controllers/cluster_controller.go:127"}
{"level":"debug","ts":"2024-05-02T14:49:10Z","msg":"Reconciliation loop end","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-example","namespace":"default"},"namespace":"default","name":"cluster-example","reconcileID":"3f432bf6-4f9c-4610-a510-6c097d46f98d","caller":"controllers/cluster_controller.go:129"}
{"level":"error","ts":"2024-05-02T14:49:10Z","msg":"Reconciler error","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"cluster-example","namespace":"default"},"namespace":"default","name":"cluster-example","reconcileID":"3f432bf6-4f9c-4610-a510-6c097d46f98d","error":"this is a fake error","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/home/leonardo/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/leonardo/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/home/leonardo/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.3/pkg/internal/controller/controller.go:227"}
[...]
```

Each reconciliation loop can be identified by the `reconcileID` fields.

**Review by leonardoce — APPROVED** [review]

**Review by sxd — APPROVED** [review]

**Comment by CiraciNicolo** [comment]
It seems clearer now 😊

**Comment by leonardoce** [comment]
I cancelled the workflow because we finished the number of workers we have.
I started it again in the EDB repo: https://github.com/EnterpriseDB/cloudnative-pg/actions/runs/9091860015

**Pull request merged**